config-service-impl/src/main/java/org/hypertrace/config/service/store/DocumentConfigStore.java

```diff
@@ -76,6 +77,56 @@ public UpsertedConfig writeConfig(
     Optional<ConfigDocument> previousConfigDoc = getLatestVersionConfigDoc(configResourceContext);
     Optional<ContextSpecificConfig> optionalPreviousConfig =
         previousConfigDoc.flatMap(this::convertToContextSpecificConfig);
+
+    Key latestDocKey = new ConfigDocumentKey(configResourceContext);
+    ConfigDocument latestConfigDocument =
+        buildConfigDocument(configResourceContext, latestConfig, userId, previousConfigDoc);
+
+    collection.upsert(latestDocKey, latestConfigDocument);
+    return optionalPreviousConfig
+        .map(previousConfig -> this.buildUpsertResult(latestConfigDocument, previousConfig))
+        .orElseGet(() -> this.buildUpsertResult(latestConfigDocument));
+  }
+
+  private List<UpsertedConfig> writeConfigs(
+      Map<ConfigResourceContext, Value> resourceContextValueMap, String userId) throws IOException {
+    Map<ConfigResourceContext, Optional<ConfigDocument>> previousConfigDocs =
+        getLatestVersionConfigDocs(resourceContextValueMap.keySet());
+    Map<Key, Document> documentsToBeUpserted = new HashMap<>();
+    List<UpsertedConfig> upsertedConfigs = new ArrayList<>();
+    for (Entry<ConfigResourceContext, Optional<ConfigDocument>> configResourceContextOptionalEntry :
+        previousConfigDocs.entrySet()) {
+      ConfigResourceContext configResourceContext = configResourceContextOptionalEntry.getKey();
+      Optional<ConfigDocument> previousConfigDoc = configResourceContextOptionalEntry.getValue();
+
+      Key latestDocKey = new ConfigDocumentKey(configResourceContext);
+      ConfigDocument latestConfigDocument =
+          buildConfigDocument(
+              configResourceContext,
+              resourceContextValueMap.get(configResourceContext),
+              userId,
+              previousConfigDoc);
+      documentsToBeUpserted.put(latestDocKey, latestConfigDocument);
+
+      upsertedConfigs.add(
```

Comment:
this hasn't actually been upserted yet has it? Can we build the results after we upsert (would also allow us to switch this to a neater streaming implementation).